[lldb] Check for abstract methods implementation in Scripted Plugin Objects

medismailben opened the pull request:
This patch enforces that every scripted object implements all the necessary abstract methods.

Every scripted affordance language interface can implement a list of abstract methods name that checked when the object is instanciated.

Since some scripting affordances implementations can be derived from template base classes, we can't check the object dictionary since it will contain the definition of the base class, so instead, this checks the scripting class dictionary.

Previously, for the various python interfaces, we used `ABC.abstractmethod` decorators but this is too language specific and doesn't work for scripting affordances that are not derived from template base classes (i.e OperatingSystem, ScriptedThreadPlan, ...), so this patch provides generic/language-agnostic checks for every scripted affordance.

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -103,24 +139,79 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    if (!checker_or_err)
+      return checker_or_err.takeError();
+
+#define LOG_ABSTRACT_METHOD_STATUS(state)                                      \
+  LLDB_LOG(GetLog(LLDBLog::Script), "Abstract method {0}.{1} " #state ".",     \
+           obj_class_name.GetString(), method_checker.first)
```
I don't think you need to do preprocessor string concatenation here.  I also don't think it's a good idea because it leads to code that looks wrong but isn't, e.g. `LOG_ABSTRACT_METHOD_STATUS(not implemented)` doesn't look like valid C++ until you look at the macro.

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -103,24 +139,79 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    for (const auto &method_checker : *checker_or_err) {
+      if (method_checker.second != AbstractMethodCheckerCases::eValid)
+        return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                       "Abstract method {0}.{1} missing ❌.");
```
Love the emoji

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -103,24 +139,79 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+
     m_object_instance_sp = StructuredData::GenericSP(
         new StructuredPythonObject(std::move(result)));
+
```
spurious whitespace

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp:
```diff
@@ -663,6 +663,18 @@ bool PythonDictionary::Check(PyObject *py_obj) {
   return PyDict_Check(py_obj);
 }
 
+bool PythonDictionary::HasKey(const llvm::Twine &key) const {
```
Why a `Twine` and not a `StringRef`?

medismailben replied:
`PythonDictionary::GetItem` have an overload that takes a `const llvm::Twine&` for the `key` parameter, so I wanted to stay consistent with the rest of the class. Also, I feel like `llvm::Twine` are offer more flexibility than `llvm::StringRef` and they offer an implicit constructor that takes a `llvm::StringRef`, so why not use it ?

bulbazord replied:
You could avoid a potential allocation with the creation of the `PythonString` entirely with a `StringRef` parameter.

Right now, line 669 invokes `key.str()` to create a `std::string` from the `key` parameter. In the worst case, this is a memory allocation just to give the `PythonString` constructor a `StringRef`. This doesn't have to be the case though, because `Twine` has a method to check if it is representable by a single `StringRef`, and if so, you can use a different method to get that `StringRef`. That would look something like this pseudocode:
```
if (key is representable as single StringRef)
  PythonString key_object(key.GetStringRefRepresentation());
 else
   PythonString key_object(key.str());
```
But  that's just more complicated. If `key` was a `StringRef` parameter, you could do `PythonString key_object(key);` and have no potential for additional allocations.

medismailben replied:
Done!

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp:
```diff
@@ -663,6 +663,18 @@ bool PythonDictionary::Check(PyObject *py_obj) {
+bool PythonDictionary::HasKey(const llvm::Twine &key) const {
+  if (!IsValid())
+    return false;
+  PythonString key_object(key.str().data());
```
`key.str()` gives you a `std::string`, `PythonString`'s constructor takes a `StringRef`. If you pass it `key.str().data()` you're building a `StringRef` from a `const char *` which means you need to do some form of `strlen`. You can construct a `StringRef` from a `std::string` directly without a `strlen`.
Assuming you don't change `key` to a `StringRef`, you can do this and save yourself from some complexity.
```suggestion
  PythonString key_object(key.str());

medismailben replied:
Good point!

bulbazord commented on lldb/test/API/functionalities/scripted_process/TestScriptedProcess.py:
```diff
@@ -60,6 +60,52 @@ def move_blueprint_to_dsym(self, blueprint_name):
         shutil.copy(blueprint_origin_path, blueprint_destination_path)
 
+    def test_missing_methods_scripted_register_context(self):
+        """Test that we only instanciate scripted processes if they implement
```
`instanciate` -> `instantiate`

bulbazord replied:
Nevermind, apparently this is also an accepted English spelling. No need to change it.

bulbazord commented on lldb/test/API/functionalities/scripted_process/missing_methods_scripted_process.py:
```diff
@@ -0,0 +1,19 @@
+
+
+def __lldb_init_module(debugger, dict):
+    if not "SKIP_SCRIPTED_PROCESS_LAUNCH" in os.environ:
```
Instead of doing this through an environment variable, is there a way we can do this with the `dict` or as a `debugger` setting?

medismailben replied:
Not really, this is how we've been doing it for all the ScriptedProcess tests.

medismailben commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,42 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+      auto callable_or_err = class_dict.GetItem(method_name);
+      if (!callable_or_err)
+        HANDLE_ERROR(method_name, AbstractMethodCheckerCases::eNotAllocated)
+      if (!PythonCallable::Check(callable_or_err.get().get()))
```
The first get's you the `lldb_private::python::PythonObject` from the `llvm::Expected`, the second gives you the `PyObject*` from the former.

JDevlieghere replied:
~So you never check the error? Doesn't this assert in a debug build?~ nvm you do on line 60.

medismailben commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedPlatformInterface.h:
```diff
@@ -26,6 +26,10 @@ class ScriptedPlatformInterface : virtual public ScriptedInterface {
     return {llvm::make_error<UnimplementedError>()};
   }
 
+  llvm::SmallVector<llvm::StringLiteral> GetAbstractMethods() const override {
+    return {};
+  }
```
If we don't implement this, we can't instantiate `ScriptedPlatformInterface` and that would be a problem for `ScriptedInterpreter::CreateScripted{Platform,Process,Thread}Interface`.

What do you think we should if we don't link against python and hence don't have a `ScriptedInterpreterPython` instance, to get each of the python interface instances ?

bulbazord replied:
This is probably a gap in my understanding, why do we instantiate `ScriptedPlatformInterface` instead of something like `ScriptedPythonPlatform`? The `Interface` at the end of `ScriptedPlatformInterface` to me feels like we shouldn't be creating objects of that type at all.

JDevlieghere replied:
I think I asked something similar in a previous PR (and forgot the answer). Give that this has come up repeatedly, it would be good to document this somewhere (for example in the relevant classes).

medismailben replied:
The same way we instantiate a base `ScriptInterpreter` instance if we don't link with python. Because we use dynamic dispatch to get the right scripting interface from the script interpreter, we cannot make any of the Scripted affordance class abstract because we won't be able to instantiate them and hence, won't be able to implement `ScriptInterpreter::CreateScripted{Platform,Process,Thread}Interface`.

Another approach, that I'm not particularly keen on doing, is making them abstract and assert in the `ScriptInterpreter::CreateScripted{Platform,Process,Thread}Interface` methods. @JDevlieghere any thoughts ?

medismailben replied:
Ok, I'll make the the base class abstract, and return a nullptr in `ScriptInterpreter::CreateScripted*Interface` in a follow-up, that should - hopefully - address all the confusion.

medismailben replied:
@bulbazord @JDevlieghere Here! https://github.com/llvm/llvm-project/pull/71465

medismailben commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,42 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    using namespace python;
+
+    std::map<llvm::StringLiteral, AbstractMethodCheckerCases> checker;
+#define HANDLE_ERROR(method_name, error)                                       \
+  {                                                                            \
+    checker[method_name] = error;                                              \
+    continue;                                                                  \
+  }
```
I feel like it helps with reducing code duplication. I'm renaming it as you suggested but lets see what other reviewers think of this.

medismailben commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -103,24 +139,79 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    if (!dict_converter.IsAllocated())
+      return llvm::createStringError(
+          llvm::inconvertibleErrorCode(),
+          "Resulting object is not a valid Python Object.");
```
I was planning on updating the error message before making the PR but I forgot 😅 Thanks for catching that!

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,27 +32,66 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    using namespace python;
+
+    std::map<llvm::StringLiteral, AbstractMethodCheckerCases> checker;
+#define SET_ERROR_AND_CONTINUE(method_name, error)                             \
+  {                                                                            \
+    checker[method_name] = error;                                              \
+    continue;                                                                  \
+  }
```
Can this be a lambda?

medismailben replied:
No because we won't be able to `continue` in the lambda.

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -99,25 +135,87 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    for (const auto &method_checker : *checker_or_err)
+      switch (method_checker.second) {
+      case AbstractMethodCheckerCases::eNotImplemented:
+        LLDB_LOG(GetLog(LLDBLog::Script),
+                 "❌ Abstract method {0}.{1} not implemented.",
+                 obj_class_name.GetString(), method_checker.first);
+        break;
+      case AbstractMethodCheckerCases::eNotAllocated:
+        LLDB_LOG(GetLog(LLDBLog::Script),
+                 "❌ Abstract method {0}.{1} not allocated.",
+                 obj_class_name.GetString(), method_checker.first);
+        break;
+      case AbstractMethodCheckerCases::eNotCallable:
+        LLDB_LOG(GetLog(LLDBLog::Script),
+                 "❌ Abstract method {0}.{1} not callable.",
+                 obj_class_name.GetString(), method_checker.first);
+        break;
+      case AbstractMethodCheckerCases::eValid:
+        LLDB_LOG(GetLog(LLDBLog::Script),
+                 "✅ Abstract method {0}.{1} implemented & valid.",
+                 obj_class_name.GetString(), method_checker.first);
```
As much as I like emojis, I'd very much prefer to keep the logs plain old ASCII.

medismailben replied:
💔

JDevlieghere reviewed (commented):
Is there any way we could avoid ever having to instantiate a `Scripted{.*}Interface`?

medismailben commented:
> Is there any way we could avoid ever having to instantiate a `Scripted{.*}Interface`?

@JDevlieghere I actually suggested something here https://github.com/llvm/llvm-project/pull/71260#discussion_r1383748012

medismailben committed:
[lldb] Check for abstract methods implementation in Scripted Plugin Objects

This patch enforces that every scripted object implements all the
necessary abstract methods.

Every scripted affordance language interface can implement a list of
abstract methods name that checked when the object is instanciated.

Since some scripting affordances implementations can be derived from
template base classes, we can't check the object dictionary since it
will contain the definition of the base class, so instead, this checks
the scripting class dictionary.

Previously, for the various python interfaces, we used `ABC.abstractmethod`
decorators but this is too language specific and doesn't work for
scripting affordances that are not derived from template base classes
(i.e OperatingSystem, ScriptedThreadPlan, ...), so this patch provides
generic/language-agnostic checks for every scripted affordance.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere reviewed (approved):
LGTM

bulbazord reviewed (approved):
I'm on board

DavidSpickett commented:
New test is failing on Windows on Arm: https://lab.llvm.org/buildbot/#/builders/219/builds/6799/steps/6/logs/stdio

The build in general was red when this initially went in, so no notification went out.

DavidSpickett commented:
Skipped: https://github.com/llvm/llvm-project/commit/ea82853499cc8780cfc86a6913d851f05d527188

I'm not 100% on what the difference between an so and a dylib is but it works on Linux as well, so I assume that a Windows test would need to build a dll to do the same thing.

medismailben commented:
@DavidSpickett Thanks for fixing the windows bot :)
